feat: add flag "output file"

Fubukimaru opened the pull request:
We now have a -o flag for the output. You can:
- Modify your file output (like what happens using .FILENAME)
- Put the output in a folder
- Both

Closes #132

Fubukimaru committed:
feat: Added -o flag

- Needs documentation
- Needs tests

Fubukimaru committed:
chore: Updated usage message

Fubukimaru committed:
chore: Added tests for issue 132 (custom output)

Fubukimaru committed:
chore: Added -o documentation

duhow commented on src/dura.y:
```diff
@@ -4579,6 +4584,13 @@ int main(int argc, char *argv[]) {
       // Make all paths relative to the main asm file
       relative_path = 1;
+    } else if (strcmp(argv[option], "-o") == 0) {
+      option += 1;
```
🐍 
```suggestion
      option++;
```

duhow commented on src/dura.y:
```diff
@@ -4593,9 +4605,9 @@ int main(int argc, char *argv[]) {
   if (fileArg == 0) {
   #if YYDEBUG == 1
-    printf("Syntax: asMSX [-z|-s|-vv|-r|-d] [file.asm]\n");
+    printf("Syntax: asMSX [-z|-s|-vv|-r|-d|-o {output file/folder}] [file.asm]\n");
```
Separa en 2 opciones, porque esto tiene un argumento adicional, en comparación al resto de opciones
```suggestion
    printf("Syntax: asMSX [-z|-s|-vv|-r|-d] [-o {output file/folder}] [file.asm]\n");
```

duhow commented on test/features/program.feature:
```diff
@@ -123,6 +123,32 @@ Feature: Test program functions
+    Given I create folder behave_test
+    Given I write the code to test.asm
+      """
+      START MAIN
```
Añade un case más:

```
.FILENAME testromname
```

Fubukimaru replied:
Done

duhow commented on test/steps/build.py:
```diff
@@ -56,9 +56,11 @@ def step_impl(context, file):
-def step_impl(context, file, flag = ""):
-    fullpath = os.path.abspath(file)
-    command = ['asmsx', flag, fullpath]
+@when('I build {file} with flag {flag} {flag_param}')
```
Purké `flag_param`? Hay problemas con `flag` al separar los argumentos?

Fubukimaru replied:
¿No se parsean por espacio? 
Así queda separado.

duhow replied:
Es que ya se separa por espacios los argumentos... flag sería "cualquier cosa" y cada palabra con espacios se separa como argumento aparte - salvo que quieras meter una "palabra con espacios" que tendría que ser tratada con comillas y algo más

duhow replied:
mm... pera.

Fubukimaru replied:
Manzana

duhow commented on test/steps/build.py:
```diff
@@ -56,9 +56,11 @@ def step_impl(context, file):
+def step_impl(context, file, flag = "", flag_param = ""):
+    # Fullpath is not needed, we always use relatives to the project
+    # fullpath = os.path.abspath(file)
+    command = ['asmsx', flag, flag_param, file] # fullpath]
```
```suggestion
    command = ['asmsx', flag, flag_param, file]
```

Fubukimaru committed:
Update src/dura.y

La pitón

Co-authored-by: David Girón <contacto@duhowpi.net>

Fubukimaru committed:
Update src/dura.y

chore: Syntax help message clearer.

Co-authored-by: David Girón <contacto@duhowpi.net>

Fubukimaru committed:
Update test/steps/build.py

chore: cleaned code comment.

Co-authored-by: David Girón <contacto@duhowpi.net>

duhow commented on test/steps/build.py:
```diff
@@ -56,9 +56,11 @@ def step_impl(context, file):
-def step_impl(context, file, flag = ""):
-    fullpath = os.path.abspath(file)
-    command = ['asmsx', flag, fullpath]
+@when('I build {file} with flag {flag} {flag_param}')
+def step_impl(context, file, flag = "", flag_param = ""):
+    # Fullpath is not needed, we always use relatives to the project
+    # fullpath = os.path.abspath(file)
+    command = ['asmsx', flag, flag_param, file]
```
```suggestion
@when('I build {file} with flag {flag}')
def step_impl(context, file, flag = ""):
    # Fullpath is not needed, we always use relatives to the project
    # fullpath = os.path.abspath(file)
    command = ['asmsx']
    if flag:
        command.extend(flag.split(" "))
    command.append(file)
```

Fubukimaru replied:
Supongo que con esto ganamos poder pasar más de un flag por ejemplo, ¿puede ser?

Fubukimaru committed:
feat: Added tests for .FILENAME pseudo-instruction

Fubukimaru committed:
Update test/steps/build.py

feat: Tests now use flag as the whole flags to be sent. With this the testing is more flexible.

Co-authored-by: David Girón <contacto@duhowpi.net>

duhow commented on test/steps/build.py:
```diff
@@ -56,9 +56,14 @@ def step_impl(context, file):
 
 @when('I build {file}')
 @when('I build {file} with flag {flag}')
+@when('I build {file} with flag {flag}')
```
duplicado
```suggestion
```

duhow committed:
remove duplicated call

jksigtermans commented:
Output file flag not functioning as expected on Ubuntu x86_64 (asmsx-linux-x86_64). 
Using: 
asmsx src/main.asm -o build/main.rom

Output:

Building MSX ROM...
-------------------------------------------------------------------------------
 asMSX v1.2.0 MSX cross-assembler. asMSX Team. [2026-02-26]
-------------------------------------------------------------------------------

Redefining output name (Overrides .FILENAME): /home/fozzie/Tmp/openmsx-project/build/main.rom
Using custom output path: /home/fozzie/Tmp/openmsx-project/build/main.rom
Fatal: cannot open /home/fozzie/Tmp/openmsx-project/build/main.rom


strace -e openat asmsx src/main.asm -o build/main.rom 
...Results in:
openat(AT_FDCWD, "/etc/ld.so.cache", O_RDONLY|O_CLOEXEC) = 3
openat(AT_FDCWD, "/lib/x86_64-linux-gnu/libm.so.6", O_RDONLY|O_CLOEXEC) = 3
openat(AT_FDCWD, "/lib/x86_64-linux-gnu/libc.so.6", O_RDONLY|O_CLOEXEC) = 3

duhow commented:
Does the folder `build` exist and has permission to write?

jksigtermans commented:
The folder exists, the permissions are: drwxrwxr-x
